File: javascript/src/model/AnsweringRuleInfo.js
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['../ApiClient', './AnsweringRuleInfoCalleeInfo', './AnsweringRuleInfoCallerInfo', './ForwardingInfo', './ScheduleInfo', './UnconditionalForwardingInfo', './VoicemailInfo'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./AnsweringRuleInfoCalleeInfo'), require('./AnsweringRuleInfoCallerInfo'), require('./ForwardingInfo'), require('./ScheduleInfo'), require('./UnconditionalForwardingInfo'), require('./VoicemailInfo'));
  } else {
    // Browser globals (root is window)
    if (!root.RingCentralApi) {
      root.RingCentralApi = {};
    }
    root.RingCentralApi.AnsweringRuleInfo = factory(root.RingCentralApi.ApiClient, root.RingCentralApi.AnsweringRuleInfoCalleeInfo, root.RingCentralApi.AnsweringRuleInfoCallerInfo, root.RingCentralApi.ForwardingInfo, root.RingCentralApi.ScheduleInfo, root.RingCentralApi.UnconditionalForwardingInfo, root.RingCentralApi.VoicemailInfo);
  }
}(this, function(ApiClient, AnsweringRuleInfoCalleeInfo, AnsweringRuleInfoCallerInfo, ForwardingInfo, ScheduleInfo, UnconditionalForwardingInfo, VoicemailInfo) {
  'use strict';

  /**
   * The AnsweringRuleInfo model module.
   * @module model/AnsweringRuleInfo
   * @version v1.0
   */

  /**
   * Constructs a new <code>AnsweringRuleInfo</code>.
   * @alias module:model/AnsweringRuleInfo
   * @class
   */
  var exports = function() {













  };

  /**
   * Constructs a <code>AnsweringRuleInfo</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/AnsweringRuleInfo} obj Optional instance to populate.
   * @return {module:model/AnsweringRuleInfo} The populated <code>AnsweringRuleInfo</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) { 
      obj = obj || new exports();

      if (data.hasOwnProperty('uri')) {
        obj['uri'] = ApiClient.convertToType(data['uri'], 'String');
      }
      if (data.hasOwnProperty('id')) {
        obj['id'] = ApiClient.convertToType(data['id'], 'String');
      }
      if (data.hasOwnProperty('type')) {
        obj['type'] = ApiClient.convertToType(data['type'], 'String');
      }
      if (data.hasOwnProperty('name')) {
        obj['name'] = ApiClient.convertToType(data['name'], 'String');
      }
      if (data.hasOwnProperty('enabled')) {
        obj['enabled'] = ApiClient.convertToType(data['enabled'], 'Boolean');
      }
      if (data.hasOwnProperty('schedule')) {
        obj['schedule'] = ScheduleInfo.constructFromObject(data['schedule']);
      }
      if (data.hasOwnProperty('calledNumbers')) {
        obj['calledNumbers'] = ApiClient.convertToType(data['calledNumbers'], [AnsweringRuleInfoCalleeInfo]);
      }
      if (data.hasOwnProperty('callers')) {
        obj['callers'] = ApiClient.convertToType(data['callers'], [AnsweringRuleInfoCallerInfo]);
      }
      if (data.hasOwnProperty('callHandlingAction')) {
        obj['callHandlingAction'] = ApiClient.convertToType(data['callHandlingAction'], 'String');
      }
      if (data.hasOwnProperty('forwarding')) {
        obj['forwarding'] = ForwardingInfo.constructFromObject(data['forwarding']);
      }
      if (data.hasOwnProperty('unconditionalForwarding')) {
        obj['unconditionalForwarding'] = UnconditionalForwardingInfo.constructFromObject(data['unconditionalForwarding']);
      }
      if (data.hasOwnProperty('voicemail')) {
        obj['voicemail'] = VoicemailInfo.constructFromObject(data['voicemail']);
      }
    }
    return obj;
  }


  /**
   * Canonical URI to the answering rule resource
   * @member {String} uri
   */
  exports.prototype['uri'] = undefined;

  /**
   * Internal identifier of an asnwering rule
   * @member {String} id
   */
  exports.prototype['id'] = undefined;

  /**
   * Type of an answering rule
   * @member {module:model/AnsweringRuleInfo.TypeEnum} type
   */
  exports.prototype['type'] = undefined;

  /**
   * Name of an answering rule specified by user
   * @member {String} name
   */
  exports.prototype['name'] = undefined;

  /**
   * Specifies if an answering rule is active or inactive
   * @member {Boolean} enabled
   */
  exports.prototype['enabled'] = undefined;

  /**
   * @member {module:model/ScheduleInfo} schedule
   */
  exports.prototype['schedule'] = undefined;

  /**
   * Answering rules are applied when calling to selected number(s)
   * @member {Array.<module:model/AnsweringRuleInfoCalleeInfo>} calledNumbers
   */
  exports.prototype['calledNumbers'] = undefined;

  /**
   * Answering rules are applied when calls are received from specified caller(s)
   * @member {Array.<module:model/AnsweringRuleInfoCallerInfo>} callers
   */
  exports.prototype['callers'] = undefined;

  /**
   * Specifies how incoming calls are forwarded
   * @member {module:model/AnsweringRuleInfo.CallHandlingActionEnum} callHandlingAction
   */
  exports.prototype['callHandlingAction'] = undefined;

  /**
   * @member {module:model/ForwardingInfo} forwarding
   */
  exports.prototype['forwarding'] = undefined;

  /**
   * @member {module:model/UnconditionalForwardingInfo} unconditionalForwarding
   */
  exports.prototype['unconditionalForwarding'] = undefined;

  /**
   * @member {module:model/VoicemailInfo} voicemail
   */
  exports.prototype['voicemail'] = undefined;


  /**
   * Allowed values for the <code>type</code> property.
   * @enum {String}
   * @readonly
   */
  exports.TypeEnum = { 
    /**
     * value: BusinessHours
     * @const
     */
    BUSINESSHOURS: "BusinessHours",
    
    /**
     * value: AfterHours
     * @const
     */
    AFTERHOURS: "AfterHours",
    
    /**
     * value: Custom
     * @const
     */
    CUSTOM: "Custom"
  };  /**
   * Allowed values for the <code>callHandlingAction</code> property.
   * @enum {String}
   * @readonly
   */
  exports.CallHandlingActionEnum = { 
    /**
     * value: ForwardCalls
     * @const
     */
    FORWARDCALLS: "ForwardCalls",
    
    /**
     * value: TakeMessagesOnly
     * @const
     */
    TAKEMESSAGESONLY: "TakeMessagesOnly",
    
    /**
     * value: PlayAnnouncementOnly
     * @const
     */
    PLAYANNOUNCEMENTONLY: "PlayAnnouncementOnly",
    
    /**
     * value: UnconditionalForwarding
     * @const
     */
    UNCONDITIONALFORWARDING: "UnconditionalForwarding"
  };

  return exports;
}));

File: csharp/src/main/csharp/IO/Swagger/Model/AnsweringRuleInfo.cs
using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace IO.Swagger.Model
{
    /// <summary>
    /// 
    /// </summary>
    [DataContract]
    public partial class AnsweringRuleInfo :  IEquatable<AnsweringRuleInfo>
    { 
    
        /// <summary>
        /// Type of an answering rule
        /// </summary>
        /// <value>Type of an answering rule</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum TypeEnum {
            
            [EnumMember(Value = "BusinessHours")]
            Businesshours,
            
            [EnumMember(Value = "AfterHours")]
            Afterhours,
            
            [EnumMember(Value = "Custom")]
            Custom
        }

    
        /// <summary>
        /// Specifies how incoming calls are forwarded
        /// </summary>
        /// <value>Specifies how incoming calls are forwarded</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum CallHandlingActionEnum {
            
            [EnumMember(Value = "ForwardCalls")]
            Forwardcalls,
            
            [EnumMember(Value = "TakeMessagesOnly")]
            Takemessagesonly,
            
            [EnumMember(Value = "PlayAnnouncementOnly")]
            Playannouncementonly,
            
            [EnumMember(Value = "UnconditionalForwarding")]
            Unconditionalforwarding
        }

    
        /// <summary>
        /// Type of an answering rule
        /// </summary>
        /// <value>Type of an answering rule</value>
        [DataMember(Name="type", EmitDefaultValue=false)]
        public TypeEnum? Type { get; set; }
    
        /// <summary>
        /// Specifies how incoming calls are forwarded
        /// </summary>
        /// <value>Specifies how incoming calls are forwarded</value>
        [DataMember(Name="callHandlingAction", EmitDefaultValue=false)]
        public CallHandlingActionEnum? CallHandlingAction { get; set; }
    
        /// <summary>
        /// Initializes a new instance of the <see cref="AnsweringRuleInfo" /> class.
        /// Initializes a new instance of the <see cref="AnsweringRuleInfo" />class.
        /// </summary>
        /// <param name="Uri">Canonical URI to the answering rule resource.</param>
        /// <param name="Id">Internal identifier of an asnwering rule.</param>
        /// <param name="Type">Type of an answering rule.</param>
        /// <param name="Name">Name of an answering rule specified by user.</param>
        /// <param name="Enabled">Specifies if an answering rule is active or inactive.</param>
        /// <param name="Schedule">Schedule.</param>
        /// <param name="CalledNumbers">Answering rules are applied when calling to selected number(s).</param>
        /// <param name="Callers">Answering rules are applied when calls are received from specified caller(s).</param>
        /// <param name="CallHandlingAction">Specifies how incoming calls are forwarded.</param>
        /// <param name="Forwarding">Forwarding.</param>
        /// <param name="UnconditionalForwarding">UnconditionalForwarding.</param>
        /// <param name="Voicemail">Voicemail.</param>

        public AnsweringRuleInfo(string Uri = null, string Id = null, TypeEnum? Type = null, string Name = null, bool? Enabled = null, ScheduleInfo Schedule = null, List<AnsweringRuleInfoCalleeInfo> CalledNumbers = null, List<AnsweringRuleInfoCallerInfo> Callers = null, CallHandlingActionEnum? CallHandlingAction = null, ForwardingInfo Forwarding = null, UnconditionalForwardingInfo UnconditionalForwarding = null, VoicemailInfo Voicemail = null)
        {
            this.Uri = Uri;
            this.Id = Id;
            this.Type = Type;
            this.Name = Name;
            this.Enabled = Enabled;
            this.Schedule = Schedule;
            this.CalledNumbers = CalledNumbers;
            this.Callers = Callers;
            this.CallHandlingAction = CallHandlingAction;
            this.Forwarding = Forwarding;
            this.UnconditionalForwarding = UnconditionalForwarding;
            this.Voicemail = Voicemail;
            
        }
        
    
        /// <summary>
        /// Canonical URI to the answering rule resource
        /// </summary>
        /// <value>Canonical URI to the answering rule resource</value>
        [DataMember(Name="uri", EmitDefaultValue=false)]
        public string Uri { get; set; }
    
        /// <summary>
        /// Internal identifier of an asnwering rule
        /// </summary>
        /// <value>Internal identifier of an asnwering rule</value>
        [DataMember(Name="id", EmitDefaultValue=false)]
        public string Id { get; set; }
    
        /// <summary>
        /// Name of an answering rule specified by user
        /// </summary>
        /// <value>Name of an answering rule specified by user</value>
        [DataMember(Name="name", EmitDefaultValue=false)]
        public string Name { get; set; }
    
        /// <summary>
        /// Specifies if an answering rule is active or inactive
        /// </summary>
        /// <value>Specifies if an answering rule is active or inactive</value>
        [DataMember(Name="enabled", EmitDefaultValue=false)]
        public bool? Enabled { get; set; }
    
        /// <summary>
        /// Gets or Sets Schedule
        /// </summary>
        [DataMember(Name="schedule", EmitDefaultValue=false)]
        public ScheduleInfo Schedule { get; set; }
    
        /// <summary>
        /// Answering rules are applied when calling to selected number(s)
        /// </summary>
        /// <value>Answering rules are applied when calling to selected number(s)</value>
        [DataMember(Name="calledNumbers", EmitDefaultValue=false)]
        public List<AnsweringRuleInfoCalleeInfo> CalledNumbers { get; set; }
    
        /// <summary>
        /// Answering rules are applied when calls are received from specified caller(s)
        /// </summary>
        /// <value>Answering rules are applied when calls are received from specified caller(s)</value>
        [DataMember(Name="callers", EmitDefaultValue=false)]
        public List<AnsweringRuleInfoCallerInfo> Callers { get; set; }
    
        /// <summary>
        /// Gets or Sets Forwarding
        /// </summary>
        [DataMember(Name="forwarding", EmitDefaultValue=false)]
        public ForwardingInfo Forwarding { get; set; }
    
        /// <summary>
        /// Gets or Sets UnconditionalForwarding
        /// </summary>
        [DataMember(Name="unconditionalForwarding", EmitDefaultValue=false)]
        public UnconditionalForwardingInfo UnconditionalForwarding { get; set; }
    
        /// <summary>
        /// Gets or Sets Voicemail
        /// </summary>
        [DataMember(Name="voicemail", EmitDefaultValue=false)]
        public VoicemailInfo Voicemail { get; set; }
    
        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.Append("class AnsweringRuleInfo {\n");
            sb.Append("  Uri: ").Append(Uri).Append("\n");
            sb.Append("  Id: ").Append(Id).Append("\n");
            sb.Append("  Type: ").Append(Type).Append("\n");
            sb.Append("  Name: ").Append(Name).Append("\n");
            sb.Append("  Enabled: ").Append(Enabled).Append("\n");
            sb.Append("  Schedule: ").Append(Schedule).Append("\n");
            sb.Append("  CalledNumbers: ").Append(CalledNumbers).Append("\n");
            sb.Append("  Callers: ").Append(Callers).Append("\n");
            sb.Append("  CallHandlingAction: ").Append(CallHandlingAction).Append("\n");
            sb.Append("  Forwarding: ").Append(Forwarding).Append("\n");
            sb.Append("  UnconditionalForwarding: ").Append(UnconditionalForwarding).Append("\n");
            sb.Append("  Voicemail: ").Append(Voicemail).Append("\n");
            
            sb.Append("}\n");
            return sb.ToString();
        }
  
        /// <summary>
        /// Returns the JSON string presentation of the object
        /// </summary>
        /// <returns>JSON string presentation of the object</returns>
        public string ToJson()
        {
            return JsonConvert.SerializeObject(this, Formatting.Indented);
        }

        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="obj">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object obj)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            return this.Equals(obj as AnsweringRuleInfo);
        }

        /// <summary>
        /// Returns true if AnsweringRuleInfo instances are equal
        /// </summary>
        /// <param name="other">Instance of AnsweringRuleInfo to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(AnsweringRuleInfo other)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            if (other == null)
                return false;

            return 
                (
                    this.Uri == other.Uri ||
                    this.Uri != null &&
                    this.Uri.Equals(other.Uri)
                ) && 
                (
                    this.Id == other.Id ||
                    this.Id != null &&
                    this.Id.Equals(other.Id)
                ) && 
                (
                    this.Type == other.Type ||
                    this.Type != null &&
                    this.Type.Equals(other.Type)
                ) && 
                (
                    this.Name == other.Name ||
                    this.Name != null &&
                    this.Name.Equals(other.Name)
                ) && 
                (
                    this.Enabled == other.Enabled ||
                    this.Enabled != null &&
                    this.Enabled.Equals(other.Enabled)
                ) && 
                (
                    this.Schedule == other.Schedule ||
                    this.Schedule != null &&
                    this.Schedule.Equals(other.Schedule)
                ) && 
                (
                    this.CalledNumbers == other.CalledNumbers ||
                    this.CalledNumbers != null &&
                    this.CalledNumbers.SequenceEqual(other.CalledNumbers)
                ) && 
                (
                    this.Callers == other.Callers ||
                    this.Callers != null &&
                    this.Callers.SequenceEqual(other.Callers)
                ) && 
                (
                    this.CallHandlingAction == other.CallHandlingAction ||
                    this.CallHandlingAction != null &&
                    this.CallHandlingAction.Equals(other.CallHandlingAction)
                ) && 
                (
                    this.Forwarding == other.Forwarding ||
                    this.Forwarding != null &&
                    this.Forwarding.Equals(other.Forwarding)
                ) && 
                (
                    this.UnconditionalForwarding == other.UnconditionalForwarding ||
                    this.UnconditionalForwarding != null &&
                    this.UnconditionalForwarding.Equals(other.UnconditionalForwarding)
                ) && 
                (
                    this.Voicemail == other.Voicemail ||
                    this.Voicemail != null &&
                    this.Voicemail.Equals(other.Voicemail)
                );
        }

        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            // credit: http://stackoverflow.com/a/263416/677735
            unchecked // Overflow is fine, just wrap
            {
                int hash = 41;
                // Suitable nullity checks etc, of course :)
                
                if (this.Uri != null)
                    hash = hash * 59 + this.Uri.GetHashCode();
                
                if (this.Id != null)
                    hash = hash * 59 + this.Id.GetHashCode();
                
                if (this.Type != null)
                    hash = hash * 59 + this.Type.GetHashCode();
                
                if (this.Name != null)
                    hash = hash * 59 + this.Name.GetHashCode();
                
                if (this.Enabled != null)
                    hash = hash * 59 + this.Enabled.GetHashCode();
                
                if (this.Schedule != null)
                    hash = hash * 59 + this.Schedule.GetHashCode();
                
                if (this.CalledNumbers != null)
                    hash = hash * 59 + this.CalledNumbers.GetHashCode();
                
                if (this.Callers != null)
                    hash = hash * 59 + this.Callers.GetHashCode();
                
                if (this.CallHandlingAction != null)
                    hash = hash * 59 + this.CallHandlingAction.GetHashCode();
                
                if (this.Forwarding != null)
                    hash = hash * 59 + this.Forwarding.GetHashCode();
                
                if (this.UnconditionalForwarding != null)
                    hash = hash * 59 + this.UnconditionalForwarding.GetHashCode();
                
                if (this.Voicemail != null)
                    hash = hash * 59 + this.Voicemail.GetHashCode();
                
                return hash;
            }
        }

    }
}
